File: jlox/src/Scanner.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    // Points to the first character in the lexeme being scanned.
    private int start = 0;
    // Points to the character currently being considered.
    private int current = 0;
    // Tracks what source line "current" is on, so we can produce tokens
    // that know their location in the source code.
    private int line = 1;

    private static final Map<String, TokenType> keywords;

    static {
        keywords = new HashMap<>();
        keywords.put("and", TokenType.AND);
        keywords.put("class", TokenType.CLASS);
        keywords.put("else", TokenType.ELSE);
        keywords.put("false", TokenType.FALSE);
        keywords.put("for", TokenType.FOR);
        keywords.put("fun", TokenType.FUN);
        keywords.put("if", TokenType.IF);
        keywords.put("nil", TokenType.NIL);
        keywords.put("or", TokenType.OR);
        keywords.put("print", TokenType.PRINT);
        keywords.put("return", TokenType.RETURN);
        keywords.put("super", TokenType.SUPER);
        keywords.put("this", TokenType.THIS);
        keywords.put("true", TokenType.TRUE);
        keywords.put("var", TokenType.VAR);
        keywords.put("while", TokenType.WHILE);
    }

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (!isAtEnd()) {
            // We are at the beginning of the next lexeme.
            start = current;
            scanToken();
        }

        tokens.add(new Token(TokenType.EOF, "", null, line));
        return tokens;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }

    private void scanToken() {
        char c = advance();
        switch (c) {
            // Single character tokens.
            case '(':
                addToken(TokenType.LEFT_PAREN);
                break;
            case ')':
                addToken(TokenType.RIGHT_PAREN);
                break;
            case '{':
                addToken(TokenType.LEFT_BRACE);
                break;
            case '}':
                addToken(TokenType.RIGHT_BRACE);
                break;
            case ',':
                addToken(TokenType.COMMA);
                break;
            case '.':
                addToken(TokenType.DOT);
                break;
            case '-':
                addToken(TokenType.MINUS);
                break;
            case '+':
                addToken(TokenType.PLUS);
                break;
            case ';':
                addToken(TokenType.SEMICOLON);
                break;
            case '*':
                addToken(TokenType.STAR);
                break;

            // One or two character tokens.
            case '!':
                // If the next character is '=', we have a '!=' token.
                // Otherwise, we have a '!' token.
                addToken(match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
                break;
            case '=':
                // If the next character is '=', we have a '==' token.
                // Otherwise, we have a '=' token.
                addToken(match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                break;
            case '<':
                // If the next character is '=', we have a '<=' token.
                // Otherwise, we have a '<' token.
                addToken(match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
                break;
            case '>':
                // If the next character is '=', we have a '>=' token.
                // Otherwise, we have a '>' token.
                addToken(match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                break;

            // "/" receives special consideration because it can be by itself,
            // which means it represents the slash token, or it can be followed
            // by another slash, which means it's the start of a comment.
            case '/':
                if (match('/')) {
                    // A comment goes until the end of the line.
                    while (peek() != '\n' && !isAtEnd()) advance();
                } else {
                    // Otherwise, we have a '/' token.
                    addToken(TokenType.SLASH);
                }
                break;

            // Whitespace characters are ignored.
            case ' ':
            case '\r':
            case '\t':
                break;

            // Newlines are tracked, so we can report the line number
            // of the token being scanned.
            case '\n':
                line++;
                break;

            // String literals.
            case '"':
                string();
                break;

            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    Lox.error(line, "Unexpected character.");
                }
                break;
        }
    }

    private char advance() {
        return source.charAt(current++);
    }

    private void addToken(TokenType type) {
        addToken(type, null);
    }

    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    // This method is like a conditional advance(),
    // it only consumes the current character if it matches what we're looking for.
    private boolean match(char expected) {
        if (isAtEnd()) return false;
        // If the next character in the source does not match the expected
        // we just peeked at it, we don't consume it and return false.
        if (source.charAt(current) != expected) return false;

        // Otherwise, we consume the character and return true.
        current++;
        return true;
    }

    private char peek() {
        if (isAtEnd()) return '\0';
        return this.source.charAt(current);
    }

    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            // We allow strings to span multiple lines.
            if (peek() == '\n') line++;
            advance();
        }

        if (isAtEnd()) {
            Lox.error(line, "Unterminated string.");
            return;
        }

        // The closing ".
        advance();

        // Trim the surrounding quotes.
        String value = source.substring(start + 1, current - 1);
        addToken(TokenType.STRING, value);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private void number() {
        while (isDigit(peek())) advance();

        // Look for a fractional part.
        // We want to do a 2-character lookahead to see if the next two
        // characters are a '.' followed by a digit. If it's not, we know
        // we have an integer and then a "dot" token.
        if (peek() == '.' && isDigit(peekNext())) {
            // Consume the "."
            advance();

            while (isDigit(peek())) advance();
        }

        addToken(TokenType.NUMBER,
                Double.parseDouble(source.substring(start, current)));
    }

    private char peekNext() {
        if (current + 1 >= source.length()) return '\0';
        return source.charAt(current + 1);
    }

    private void identifier() {
        while (isAlphaNumeric(peek())) advance();

        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null) type = TokenType.IDENTIFIER;
        addToken(type);
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                c == '_';
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }
}

File: dlox/lib/scanner.dart
import 'lox.dart';
import 'token.dart';
import 'token_type.dart';

class Scanner {
  final String _source;
  final List<Token> _tokens = [];

  int _start = 0;
  int _current = 0;
  int _line = 1;
  int _column = 0;

  // reserved keywords
  static final Map<String, TokenType> _keywords = {
    "and": TokenType.keywordAnd,
    "class": TokenType.keywordClass,
    "else": TokenType.keywordElse,
    "false": TokenType.keywordFalse,
    "for": TokenType.keywordFor,
    "fun": TokenType.keywordFun,
    "if": TokenType.keywordIf,
    "nil": TokenType.keywordNil,
    "or": TokenType.keywordOr,
    "print": TokenType.keywordPrint,
    "return": TokenType.keywordReturn,
    "super": TokenType.keywordSuper,
    "this": TokenType.keywordThis,
    "true": TokenType.keywordTrue,
    "var": TokenType.keywordVar,
    "while": TokenType.keywordWhile,
    "continue": TokenType.keywordContinue,
  };

  Scanner(this._source);

  List<Token> scanTokens() {
    while (!_isAtEnd()) {
      // We are at the beginning of the next lexeme.
      _start = _current;
      _scanToken();
    }

    _tokens.add(Token(
      type: TokenType.eof,
      lexeme: "",
      literal: null,
      line: _line,
      column: _column,
    ));
    return _tokens;
  }

  void _scanToken() {
    final c = _advance();
    switch (c) {
      case '(':
        _addToken(TokenType.leftParen);
      case ')':
        _addToken(TokenType.rightParen);
      case '{':
        _addToken(TokenType.leftBrace);
      case '}':
        _addToken(TokenType.rightBrace);
      case ',':
        _addToken(TokenType.comma);
      case '.':
        _addToken(TokenType.dot);
      case '-':
        _addToken(TokenType.minus);
      case '+':
        _addToken(TokenType.plus);
      case ';':
        _addToken(TokenType.semicolon);
      case '*':
        _addToken(TokenType.star);
      case '!':
        _addToken(_match('=') ? TokenType.bangEqual : TokenType.bang);
      case '=':
        _addToken(_match('=') ? TokenType.equalEqual : TokenType.equal);
      case '<':
        _addToken(_match('=') ? TokenType.lessEqual : TokenType.less);
      case '>':
        _addToken(_match('=') ? TokenType.greaterEqual : TokenType.greater);
      case '/':
        if (_match('/')) {
          // A comment goes until the end of the line.
          while (_peek() != '\n' && !_isAtEnd()) {
            _advance();
          }
        } else {
          _addToken(TokenType.slash);
        }
      case ' ' || '\r' || '\t':
        // Ignore whitespace.
        break;
      case '\n':
        _line++;
        _column = 0;
      case '"':
        _string();
      default:
        if (_isDigit(c)) {
          _number();
        } else if (_isAlpha(c)) {
          _identifier();
        } else {
          Lox.error(_line, "Unexpected character.");
        }
    }
  }

  String _advance() {
    _current++;
    final currentString = _source[_current - 1];
    if (currentString == '\n') {
      _column = 0;
    } else {
      _column++;
    }
    return currentString;
  }

  /// Adds a token to the list of tokens.
  void _addToken(TokenType type, [Object? literal, int? line, int? column]) {
    line ??= _line;
    String text = _source.substring(_start, _current);
    // The column is the current column, minus the length of the text,
    // unless specified.
    column ??= (_column - text.length) + 1;

    _tokens.add(Token(
      type: type,
      lexeme: text,
      literal: literal,
      line: line,
      column: column,
    ));
  }

  /// Checks if the character at [current] matches the received
  /// [character]. If it does, it advances [current] and returns
  /// true. Otherwise, it returns false without advancing.
  bool _match(String character) {
    if (_isAtEnd()) return false;
    if (_source[_current] != character) return false;

    _column++;
    _current++;
    return true;
  }

  /// Whether we've reached the end of the source code.
  bool _isAtEnd() {
    return _current >= _source.length;
  }

  String? _peek() {
    if (_isAtEnd()) return null;
    return _source[_current];
  }

  void _string() {
    // keep track of where the line and column started
    int line = _line;
    int column = _column;
    while (_peek() != '"' && !_isAtEnd()) {
      if (_peek() == '\n') {
        _line++;
      }
      _advance();
    }

    // Unterminated string.
    if (_isAtEnd()) {
      Lox.error(_line, "Unterminated string.");
      return;
    }

    // The closing ".
    _advance();

    // Trim the surrounding quotes.
    String value = _source.substring(_start + 1, _current - 1);

    _addToken(TokenType.string, value, line, column);
  }

  bool _isDigit(String? character) {
    if (character == null) return false;
    return character.compareTo('0') >= 0 && character.compareTo('9') <= 0;
  }

  void _number() {
    while (_isDigit(_peek())) {
      _advance();
    }

    // Look for a fractional part.
    if (_peek() == '.' && _isDigit(_peekNext())) {
      // Consume the "."
      _advance();

      while (_isDigit(_peek())) {
        _advance();
      }
    }

    _addToken(
      TokenType.number,
      double.parse(_source.substring(_start, _current)),
    );
  }

  String? _peekNext() {
    if (_current + 1 >= _source.length) return null;
    return _source[_current + 1];
  }

  bool _isAlpha(String character) {
    return (character.compareTo('a') >= 0 && character.compareTo('z') <= 0) ||
        (character.compareTo('A') >= 0 && character.compareTo('Z') <= 0) ||
        character == '_';
  }

  void _identifier() {
    while (_isAlphaNumeric(_peek())) {
      _advance();
    }

    // See if the identifier is a reserved word.
    String text = _source.substring(_start, _current);
    TokenType? type = _keywords[text];
    type ??= TokenType.identifier;
    _addToken(type, text);
  }

  bool _isAlphaNumeric(String? peek) {
    if (peek == null) return false;
    return _isAlpha(peek) || _isDigit(peek);
  }
}
